### ISSUE TITLE
coroutine-annotated generator should be  `AwaitableGenerator`

### ISSUE THREAD

**Issue by rsokl** [microsoft/pyright#4571]
pyright treats a generator decorated with `types.coroutine` as a function that returns an `Awaitable`, whereas mypy sees that it returns an `AwaitableGenerator`. This leads to false positives when coroutine is utilized as a generator.

```python
import types
from typing import Generator

@types.coroutine
def sleep() -> Generator[int, str, bool]:
    actual = yield 1
    return actual.isalnum()


# pyright: Awaitable[bool]
# mypy: AwaitableGenerator[int, str, bool, Generator[int, str, bool]]
reveal_type(sleep())

# pyright: error Awaitable[bool]" is incompatible with protocol "SupportsNext[_T@next]
# mypy: int
reveal_type(sleep()) reveal_type(next(sleep()))

# pyright: Cannot access member "send" for type "Awaitable[bool]
# mypy: int
reveal_type(sleep().send("a"))
```

**Comment by erictraut** [comment]
Pyright is correctly applying the type definition for the `typing.coroutine` as it appears in the typeshed stub `types.pyi`, so I don't consider this a bug in pyright.

There is a comment in the stub that indicates that the correct type is not possible without the use of an intersection, which isn't currently supported in the Python type system.

```python
# it's not really an Awaitable, but can be used in an await expression. Real type: Generator & Awaitable
# The type: ignore is due to overlapping overloads, not the use of ParamSpec
@overload
def coroutine(func: Callable[_P, Generator[Any, Any, _R]]) -> Callable[_P, Awaitable[_R]]: ...  # type: ignore[misc]
@overload
def coroutine(func: _Fn) -> _Fn: ...
```

You could inquire in the typeshed project whether there's a better way for the type stub to define this.

I suspect that mypy is special casing this in the code logic, thus overriding the type definition in the typeshed stub. That's something we generally don't do in pyright unless it's a very common case. Overriding behaviors in typeshed inevitably leads to maintenance problems and broken edge cases. I'm going to close this for now. We might consider reopening it and adding special logic for this case in the future if there's sufficient signal from pyright users (in the form of upvotes).
